### PR TITLE
Add capability change compaction planner

### PR DESCRIPTION
**Pull request by jwilder**
This adds the capability to use a different compaction planner programatically.

### CODE REVIEW

File: tsdb/engine/tsm1/compact.go [hunk 1]
```diff
@@ -1279,21 +1300,26 @@ func (a blocks) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
...
 // size indicates the maximum number of values to encode in a single block.
-func NewTSMKeyIterator(size int, fast bool, interrupt chan struct{}, readers ...*TSMReader) (KeyIterator, error) {
+func NewTSMKeyIterator(size int, fast bool, filter tsdb.CompactionFilter, interrupt chan struct{}, readers ...*TSMReader) (KeyIterator, error) {
 	var iter []*BlockIterator
 	for _, r := range readers {
 		iter = append(iter, r.BlockIterator())
 	}
 
+	if filter == nil {
+		filter = &noFilter{}
```
Comment (stuartcarnie) [inline tsdb/engine/tsm1/compact.go]: Can you make this a static var, so it doesn't incur any allocations

```go
var defaultFilter tsdb.CompactionFilter = &noFilter{}
```

File: tsdb/engine/tsm1/compact.go [hunk 2]
```diff
@@ -661,6 +661,19 @@ func (c *DefaultPlanner) Release(groups []CompactionGroup) {
...
 
+// noFilter is a CompactionFilter that does not filter anything
+type noFilter struct {
+}
+
+func (f *noFilter) FilterBlock(key []byte, minTime, maxTime int64) bool {
+	return false
+}
+
+func (f *noFilter) FilterTimeRange(key []byte) (int64, int64) {
```
Comment (stuartcarnie) [inline tsdb/engine/tsm1/compact.go]: Is it worth moving `tsm1.TimeRange` to the `tsdb` package and returning that value here. By doing so, we could add a concept like `IsZero() bool` or `IsEmpty() bool` that can be used elsewhere, such as:

```go

// Remove any values filtered globally
filterRange := k.compactionFilter.FilterTimeRange(k.blocks[i].key)
if !filterRange.IsZero() {
    v = IntegerValues(v).Exclude(filterRange.Min, filterRange.Max)
    // or better still:
    v = IntegerValues(v).Exclude(filterRange)
}
```

Comment (jwilder) [reply tsdb/engine/tsm1/compact.go]: Maybe. I'm going to hold off of that change for now.

Comment (stuartcarnie) [reply tsdb/engine/tsm1/compact.go]: I am happy to take than on later 👍

**Comment by jwilder** [comment]
I removed the filtering part of this PR.  That can lead to problems where the index is inconsistent w/ TSM data.  I've left the hooks to change the compaction planner though.

**Pull request merged**